### PR TITLE
[SDK-3808] Optionally sign the session store cookie

### PR DESCRIPTION
**Pull request by adamjmcgrath**
### Description

Add an option `signSessionStoreCookie` (default `false`) - when true and using a custom session store, the session store cookie will be signed. e.g.

```js
// session store => {"123abc": { ... session ... } }

// before
Cookie: appSession=abc123

// after (with cryptographically secure signature)
Cookie: appSession=abc123.b1825178d1f21747e60d329c2cfc21a9c
```

The default session id is a cryptographically secure random value, but this is useful when you want to override `genid` and you are only generating a unique value, not a cryptographically secure random value.

Also adding a `requireSignedSessionStoreCookie` config option to allow zero down time migration to signed session store cookies.

fafa0e0f88f7b84a88161b369609200459a02157 The first commit is just moving some code into another file to make it easier to share
d1c3dc0e84ae87cefb2fc420474a292c99f55e33 This is the main commit that implements signed session store cookies

### References
[express-session docs](https://www.npmjs.com/package/express-session#:~:text=Using%20a%20secret%20that%20cannot%20be%20guessed%20will%20reduce%20the%20ability%20to%20hijack%20a%20session%20to%20only%20guessing%20the%20session%20ID%20(as%20determined%20by%20the%20genid%20option).)
[genid](https://auth0.github.io/express-openid-connect/interfaces/SessionConfigParams.html#genid)

### Testing

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used, if not the default branch

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -637,12 +640,38 @@ interface SessionConfigParams {
...
+   * cryptographically strong random value of sufficient size.
+   */
+  signSessionStoreCookie: boolean;
+
+  /**
+   * If you enable {@Link signSessionStoreCookie} your existing sessions will
+   * be invalidated. You can use this flag to temporarily allow unsigned cookies
+   * while you sign your user's session cookies. For example:
+   *
+   * Set {@Link signSessionStoreCookie} to `true` and {@Link requireSignedSessionStoreCookie} to `false`
```
Comment (Widcket) [inline index.d.ts:665]:
```suggestion
   * Set {@Link signSessionStoreCookie} to `true` and {@Link requireSignedSessionStoreCookie} to `false`.
```

File: index.d.ts [hunk 2]
```diff
@@ -637,12 +640,38 @@ interface SessionConfigParams {
...
+   * be invalidated. You can use this flag to temporarily allow unsigned cookies
+   * while you sign your user's session cookies. For example:
+   *
+   * Set {@Link signSessionStoreCookie} to `true` and {@Link requireSignedSessionStoreCookie} to `false`.
+   * Wait for your {@Link rollingDuration} (default 1 day) or {@Link absoluteDuration} (default 1 week)
+   * to pass (which ever comes first). By this time all your sessions cookies will either be signed or
+   * have expired, then you can remove the {@Link requireSignedSessionStoreCookie} config option which
+   * will set it to `true`.
+   *
+   * Signed session store cookies will be mandatory in the next Major release.
```
Comment (Widcket) [inline index.d.ts:671]:
```suggestion
   * Signed session store cookies will be mandatory in the next major release.
```

File: lib/crypto.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+const ENCRYPTION_INFO = 'JWE CEK';
+const SIGNING_INFO = 'JWS Cookie Signing';
+const DIGEST = 'sha256';
+
+let encryption, signing;
+
+/**
+ *
+ * Derives appropriate sized keys from the end-user provided secret random string/passphrase using
+ * HKDF (HMAC-based Extract-and-Expand Key Derivation Function) defined in RFC 8569
```
Comment (Widcket) [inline lib/crypto.js:14]:
```suggestion
 * HKDF (HMAC-based Extract-and-Expand Key Derivation Function) defined in RFC 8569.
```

File: lib/crypto.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+const generateSignature = (cookie, value, key) => {
+  const payload = getPayload(cookie, value);
+  return JWS.sign.flattened(payload, key, header).signature;
+};
+const verifySignature = (cookie, value, signature, keystore) => {
+  try {
+    return !!JWS.verify(
+      flattenedJWSFromCookie(cookie, value, signature),
+      keystore,
+      { algorithm: 'HS256', crit: ['b64'] }
```
Comment (Widcket) [inline lib/crypto.js:86]: Any way to not duplicate these values that are already used in the `header` constant?

Comment (adamjmcgrath) [reply lib/crypto.js]: 👍 have moved these into constants 

Also, noticed a type - thanks

**Review by Widcket — APPROVED** [review]

**Pull request merged**